# RQ4


In [1]:
import json
import pandas as pd

In [4]:
Italy = pd.read_json(r"C:\Users\Nino\Desktop\events_Italy.json")
Germany = pd.read_json(r"C:\Users\Nino\Desktop\events_Germany.json")
France = pd.read_json(r"C:\Users\Nino\Desktop\events_France.json")
England = pd.read_json(r"C:\Users\Nino\Desktop\events_England.json")
Spain = pd.read_json(r"C:\Users\Nino\Desktop\events_Spain.json")

In [5]:
events = pd.concat([Italy,Germany,France,England,Spain], axis=0)

In [6]:
events

eventId           eventName     eventSec         id  matchId  \
0             8                Pass     2.530536  180423957  2575959   
1             8                Pass     3.768418  180423958  2575959   
2             7  Others on the ball     4.868265  180423959  2575959   
3             1                Duel     8.114676  180423960  2575959   
4             1                Duel     8.647892  180423961  2575959   
5             8                Pass    10.376066  180423962  2575959   
6             1                Duel    10.796797  180423963  2575959   
7             1                Duel    11.527556  180423976  2575959   
8             1                Duel    12.156579  180423977  2575959   
9             1                Duel    12.418937  180423965  2575959   
10            8                Pass    16.240563  180423979  2575959   
11            7  Others on the ball    17.251513  180423982  2575959   
12            1                Duel    17.715422  180423967  2575959   
13            1                Duel    17.791970  180423983  2575959   
14            8                Pass    19.153235  180423968  2575959   
15            8                Pass    20.873026  180423969  2575959   
16            8                Pass    21.503835  180423970  2575959   
17            8                Pass    23.201103  180423971  2575959   
18            1                Duel    26.751844  180423972  2575959   
19            1                Duel    26.752540  180423984  2575959   
20            8                Pass    27.700356  180423973  2575959   
21            1                Duel    28.483430  180423974  2575959   
22            1                Duel    29.133186  180423985  2575959   
23            1                Duel    30.060792  180423975  2575959   
24            1                Duel    31.352912  180423986  2575959   
25            2                Foul    33.869510  180423987  2575959   
26            3           Free Kick    52.956101  180423978  2575959   
27            1                Duel    55.973765  180423980  2575959   
28            1                Duel    56.176830  180423992  2575959   
29            8                Pass    57.313039  180423981  2575959   
...         ...                 ...          ...        ...      ...   
628629        1                Duel  2826.155800  253302636  2565927   
628630        1                Duel  2826.959643  253302680  2565927   
628631        7  Others on the ball  2829.338423  253302682  2565927   
628632        8                Pass  2830.558379  253302683  2565927   
628633        8                Pass  2832.161413  253302685  2565927   
628634        7  Others on the ball  2832.512763  253302644  2565927   
628635        5        Interruption  2838.422522  253302645  2565927   
628636        3           Free Kick  2853.997053  253302703  2565927   
628637       10                Shot  2857.346465  253302686  2565927   
628638        5        Interruption  2857.454243  253302647  2565927   
628639        3           Free Kick  2873.020476  253302649  2565927   
628640        8                Pass  2874.769542  253302650  2565927   
628641        8                Pass  2876.959208  253302651  2565927   
628642        1                Duel  2880.014550  253302653  2565927   
628643        1                Duel  2880.551604  253302688  2565927   
628644        1                Duel  2880.965159  253302656  2565927   
628645        1                Duel  2881.224312  253302689  2565927   
628646        1                Duel  2882.558967  253302657  2565927   
628647        1                Duel  2883.225650  253302691  2565927   
628648        2                Foul  2885.241461  253302659  2565927   
628649        3           Free Kick  2917.139970  253302692  2565927   
628650        8                Pass  2925.547631  253302662  2565927   
628651        7  Others on the ball  2927.064747  253302664  2565927   
628652        8                Pass  2931.782904

since in this RQ we are just interested in "pass" events, we filter the rows that has eventId=8

In [7]:
pass_events = events[events.eventId == 8]
pass_events

eventId eventName     eventSec         id  matchId matchPeriod  \
0             8      Pass     2.530536  180423957  2575959          1H   
1             8      Pass     3.768418  180423958  2575959          1H   
5             8      Pass    10.376066  180423962  2575959          1H   
10            8      Pass    16.240563  180423979  2575959          1H   
14            8      Pass    19.153235  180423968  2575959          1H   
15            8      Pass    20.873026  180423969  2575959          1H   
16            8      Pass    21.503835  180423970  2575959          1H   
17            8      Pass    23.201103  180423971  2575959          1H   
20            8      Pass    27.700356  180423973  2575959          1H   
29            8      Pass    57.313039  180423981  2575959          1H   
34            8      Pass    94.813038  180423997  2575959          1H   
35            8      Pass    97.794281  180423998  2575959          1H   
36            8      Pass    99.324654  180424001  2575959          1H   
37            8      Pass   101.978336  180424000  2575959          1H   
41            8      Pass   110.967192  180424007  2575959          1H   
42            8      Pass   116.701879  180424008  2575959          1H   
43            8      Pass   117.986911  180424009  2575959          1H   
44            8      Pass   118.372288  180424010  2575959          1H   
53            8      Pass   160.255226  180424015  2575959          1H   
55            8      Pass   163.195949  180424016  2575959          1H   
56            8      Pass   167.259232  180424017  2575959          1H   
57            8      Pass   173.754622  180424018  2575959          1H   
58            8      Pass   175.173176  180424019  2575959          1H   
59            8      Pass   178.127000  180424020  2575959          1H   
60            8      Pass   181.969441  180424022  2575959          1H   
61            8      Pass   185.706491  180424023  2575959          1H   
62            8      Pass   188.436954  180424025  2575959          1H   
70            8      Pass   197.987393  180424037  2575959          1H   
71            8      Pass   201.317410  180424039  2575959          1H   
72            8      Pass   203.034225  180424040  2575959          1H   
...         ...       ...          ...        ...      ...         ...   
628579        8      Pass  2690.093773  253302615  2565927          2H   
628581        8      Pass  2693.158074  253302616  2565927          2H   
628583        8      Pass  2699.050573  253302624  2565927          2H   
628586        8      Pass  2700.781421  253302597  2565927          2H   
628587        8      Pass  2701.561538  253302598  2565927          2H   
628588        8      Pass  2703.207109  253302599  2565927          2H   
628589        8      Pass  2708.283680  253302601  2565927          2H   
628590        8      Pass  2711.313988  253302634  2565927          2H   
628592        8      Pass  2713.765334  253302635  2565927          2H   
628595        8      Pass  2721.203942  253302639  2565927          2H   
628613        8      Pass  2776.189218  253302623  2565927          2H   
628614        8      Pass  2782.099866  253302625  2565927          2H   
628616        8      Pass  2787.630183  253302628  2565927          2H   
628617        8      Pass  2789.455395  253302629  2565927          2H   
628618        8      Pass  2791.610988  253302630  2565927          2H   
628623        8      Pass  2808.463744  253302669  2565927          2H   
628624        8      Pass  2813.117759  253302672  2565927          2H   
628625        8      Pass  2817.686043  253302675  2565927          2H   
628626        8      Pass  2820.002123  253302677  2565927          2H   
628627        8      Pass  2824.595914  253302678  2565927          2H   
628628        8      Pass  2825.590369  253302679  2565927          2H   
628632        8      Pass  2830.558379  253302683  2565927          2H   
628633      

Now we want to know about accurate passes so we filter rows that contain tag id 1801

In [9]:
acc_passes = pass_events[pass_events['tags'].apply(lambda x: pd.Series([dict['id'] for dict in x]).eq(1801).any())]

In [10]:
acc_passes

eventId eventName     eventSec         id  matchId matchPeriod  \
0             8      Pass     2.530536  180423957  2575959          1H   
1             8      Pass     3.768418  180423958  2575959          1H   
5             8      Pass    10.376066  180423962  2575959          1H   
10            8      Pass    16.240563  180423979  2575959          1H   
14            8      Pass    19.153235  180423968  2575959          1H   
15            8      Pass    20.873026  180423969  2575959          1H   
16            8      Pass    21.503835  180423970  2575959          1H   
17            8      Pass    23.201103  180423971  2575959          1H   
20            8      Pass    27.700356  180423973  2575959          1H   
34            8      Pass    94.813038  180423997  2575959          1H   
35            8      Pass    97.794281  180423998  2575959          1H   
36            8      Pass    99.324654  180424001  2575959          1H   
41            8      Pass   110.967192  180424007  2575959          1H   
42            8      Pass   116.701879  180424008  2575959          1H   
43            8      Pass   117.986911  180424009  2575959          1H   
53            8      Pass   160.255226  180424015  2575959          1H   
55            8      Pass   163.195949  180424016  2575959          1H   
57            8      Pass   173.754622  180424018  2575959          1H   
58            8      Pass   175.173176  180424019  2575959          1H   
59            8      Pass   178.127000  180424020  2575959          1H   
60            8      Pass   181.969441  180424022  2575959          1H   
61            8      Pass   185.706491  180424023  2575959          1H   
62            8      Pass   188.436954  180424025  2575959          1H   
70            8      Pass   197.987393  180424037  2575959          1H   
71            8      Pass   201.317410  180424039  2575959          1H   
74            8      Pass   207.542734  180424045  2575959          1H   
75            8      Pass   210.057782  180424047  2575959          1H   
76            8      Pass   213.785432  180424049  2575959          1H   
84            8      Pass   237.518669  180424061  2575959          1H   
85            8      Pass   238.671690  180424064  2575959          1H   
...         ...       ...          ...        ...      ...         ...   
628561        8      Pass  2579.699474  253302587  2565927          2H   
628562        8      Pass  2586.089687  253302589  2565927          2H   
628563        8      Pass  2589.275233  253302590  2565927          2H   
628564        8      Pass  2593.265012  253302592  2565927          2H   
628573        8      Pass  2676.889159  253302610  2565927          2H   
628579        8      Pass  2690.093773  253302615  2565927          2H   
628581        8      Pass  2693.158074  253302616  2565927          2H   
628583        8      Pass  2699.050573  253302624  2565927          2H   
628586        8      Pass  2700.781421  253302597  2565927          2H   
628587        8      Pass  2701.561538  253302598  2565927          2H   
628588        8      Pass  2703.207109  253302599  2565927          2H   
628595        8      Pass  2721.203942  253302639  2565927          2H   
628613        8      Pass  2776.189218  253302623  2565927          2H   
628614        8      Pass  2782.099866  253302625  2565927          2H   
628616        8      Pass  2787.630183  253302628  2565927          2H   
628617        8      Pass  2789.455395  253302629  2565927          2H   
628618        8      Pass  2791.610988  253302630  2565927          2H   
628623        8      Pass  2808.463744  253302669  2565927          2H   
628624        8      Pass  2813.117759  253302672  2565927          2H   
628625        8      Pass  2817.686043  253302675  2565927          2H   
628626        8      Pass  2820.002123  253302677  2565927          2H   
628627        8      Pass  2824.595914  253302678  2565927          2H   
628628      

In [15]:
ratio = ( acc_passes["playerId"].value_counts() / pass_events["playerId"].value_counts() ).sort_values(ascending=[False])
ratio

568583    1.000000
257752    1.000000
22053     1.000000
472696    1.000000
472194    1.000000
400564    1.000000
20972     1.000000
415747    1.000000
461671    1.000000
286823    1.000000
15254     1.000000
291528    1.000000
329026    1.000000
331913    1.000000
337656    1.000000
339812    1.000000
342088    1.000000
447538    1.000000
447254    1.000000
352985    1.000000
353241    1.000000
354470    1.000000
357992    1.000000
360778    1.000000
364640    1.000000
364641    1.000000
423162    1.000000
397030    1.000000
367391    1.000000
391551    1.000000
            ...   
375771    0.541667
399514    0.529412
99437     0.504202
8566      0.500000
15201     0.500000
272436    0.500000
294697    0.500000
8752      0.500000
426704    0.500000
13745     0.500000
37959     0.500000
279721    0.500000
365707    0.500000
301016    0.500000
26378     0.500000
18728     0.500000
26006     0.473684
404285    0.470588
6363      0.470588
272234    0.454545
248632    0.428571
15095     0.

As you can see some of our rows have value 1. lets evaluate one of this points :

In [18]:
print(acc_passes["playerId" ].value_counts()[568583])
print(pass_events["playerId" ].value_counts()[568583])

3
3


This player had only 3 passes and all of them were accurate so we can't judge whether he is a good palyer or not based on this information. Now we put a threshhold to solve this problem. for setting a threshold, it's good to have an initial understanding of our data:

In [20]:
pass_events["playerId"].value_counts().describe()

count    2565.000000
mean      610.275244
std       524.822133
min         1.000000
25%       161.000000
50%       496.000000
75%       926.000000
max      3118.000000
Name: playerId, dtype: float64

As you can see the first quartile is 161 and it means that 75% of our population had more than 161 passes. I put this number as threshold :